### ISSUE TITLE
cmd/go: document build constraints in 'go help'

### ISSUE THREAD

**Issue by pjweinb** [golang/go#37018]
### What version of Go are you using (`go version`)?

<pre>
go version devel +a50c3ffbd4 linux/amd64
</pre>

### What operating system and processor architecture are you using (`go env`)?

<details><summary><code>go env</code> Output</summary><br><pre>
GOARCH="amd64"
GOBIN=""
GOCACHE="/usr/local/google/home/pjw/.cache/go-build"
GOENV="/usr/local/google/home/pjw/.config/go/env"
GOEXE=""
GOFLAGS=""
GOHOSTARCH="amd64"
GOHOSTOS="linux"
GOINSECURE=""
GONOPROXY=""
GONOSUMDB=""
GOOS="linux"

</pre></details>

### What did you do?
In x/tools/internal/lsp/span/uri_windows_test.go I removed // +build window, and renamed the test function to TestURIA (to avoid conflicts with uri_test.go)

I then ran go test -run URI .
### What did you expect to see?
I expected to see it run TestURI (from uri_test.go) and TestURIA (from uri_windows_test.go)

### What did you see instead?
It just ran TestURI

[If I rename uri_window_test.go then go test will run TestURIA]

**Comment by ALTree** [comment]
The build tag in the filename is preventing the test from being run.

**Comment by pjweinbgo** [comment]
It would be nice if that were documented somewhere in ordinary help, rather than hoping users will paw through the source code.

**Comment by ALTree** [comment]
This is documented. From https://golang.org/pkg/go/build

```
If a file's name, after stripping the extension and a possible _test suffix, 
matches any of the following patterns:

*_GOOS
*_GOARCH
*_GOOS_GOARCH

(example: source_windows_amd64.go) where GOOS and GOARCH 
represent any known operating system and architecture values 
respectively, then the file is considered to have an implicit build 
constraint requiring those terms (in addition to any explicit 
constraints in the file). 
```

**Comment by jayconrod** [comment]
> It would be nice if that were documented somewhere in ordinary help, rather than hoping users will paw through the source code.

I agree with this. Build constraints are mentioned a few times in https://golang.org/cmd/go/, but they aren't defined. There's only one reference to the `go/build` package documentation under `go help filetype`. It's pretty hard to find.

I think we should add a `go help build-constraints` page that describes tags and suffixes more thoroughly, and it should be referenced from any section that mentions build constraints.

**Comment by ALTree** [comment]
@jayconrod 

> There's only one reference to the go/build package documentation under go help filetype

This is not true; the feature is pretty clearly documented in the https://golang.org/pkg/go/build section I pasted above.

**Comment by jayconrod** [comment]
Very true, but I don't think it's enough. If you're reading `go help` pages, it's not currently clear that you need to read the docs for `go/build`. `go/build` is only referenced twice in the context of build constraints: once in the description for the `-tags` flag, and again at the end of `go help filetype`. It's not obvious in either case that GOOS / GOARCH suffixes are meaningful.

**Comment by mlowicki** [comment]
@jayconrod I can work on this one. I'll add `go help build-constraints`. It seems that content there will be pretty much the same as in https://golang.org/pkg/go/build/#hdr-Build_Constraints. Should I replace https://golang.org/pkg/go/build/#hdr-Build_Constraints with reference to new page (`See 'go help build-constraints'`) to avoid duplication?

**Comment by jayconrod** [comment]
@mlowicki Thanks for looking at this.

I don't think the documentation in `go/build` should be replaced. Documentation in `go help` should probably just be a quick reference with pointers to more in-depth HTML documentation. The `go/build` documentation is the closest we have right now to the latter, but we may want to expand it somewhere else in the future.

**Comment by jayconrod** [comment]
Reopening this. @rsc pointed out there's no need to document build constraints in two places. `go help buildconstraint` is almost as long as the section in `go/build`. The `go help` page is more discoverable, so we should make that version complete and canonical and just point there from `go/build`.

I'll send a CL that does this shortly.
